src/extension/tab/hook.ts

```diff
@@ -54,27 +55,39 @@ const tab = createWindowActor<ClientMessage>(window);
 const messageAdapter = createWindowMessageAdapter(window);
 const handleRpc = createRpcHandler<DevtoolsRPCMessage>(messageAdapter);
 const rpcClient = createRpcClient<ErrorCodesHandler>(messageAdapter);
-const knownClients = new Set<ApolloClient<any>>();
+
+function getQueriesForClient(client: ApolloClient<unknown> | undefined) {
+  const ac = getPrivateAccess(client);
+  if (ac?.queryManager.getObservableQueries) {
+    return getQueries(ac.queryManager.getObservableQueries("active"));
+  } else {
+    return getQueriesLegacy(ac?.queryManager["queries"]);
+  }
+}
+
+function getMutationsForClient(client: ApolloClient<unknown> | undefined) {
+  const ac = getPrivateAccess(client);
+
+  return getMutations(
+    (ac?.queryManager.mutationStore?.getStore
+      ? // @ts-expect-error Apollo Client 3.0 - 3.2
+        ac.queryManager.mutationStore?.getStore()
+      : // Apollo Client 3.3
+        ac?.queryManager.mutationStore) ?? {}
+  );
+}
+
+// Keep a reverse mapping of client -> id to ensure we don't register the same
+// client multiple times.
+const knownClients = new Map<ApolloClient<SafeAny>, string>();
```

Comment:
```suggestion
const knownClients = new WeakMap<ApolloClient<SafeAny>, string>();
```

Let's ensure that the devtools don't cause any memory leaks

Reply:
Unfortunately I need to be able to traverse this map to be able to look up which client belongs to which `id`, otherwise it defeats the purpose of this feature.

I probably could use some combination of `WeakRef` and `WeakSet`/`WeakMap` for this to be able to perform some kind of iteration along with `FinalizationRegistry` to detect when the client has been garbage collected, but definitely a bit more complicated than it seems. How do you feel about a followup PR that does this in isolation?